### PR TITLE
3.3 tag cross thread

### LINKED ISSUE #12656: copy dubbo.tag cross thread
- [ - ] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [ - ] I have searched the [release notes](https://github.com/apache/dubbo/releases) of this repository and believe that this is not a duplicate.

## Describe the feature
suggest add annotation to copy dubbo.tag cross thread , 
like my commit in skywalking: 
https://github.com/apache/skywalking-java/commit/8612f80260c8dd691a5c8308625fae5fca03ade2

### PR DESCRIPTION
**Pull request by carlvine500**
## What is the purpose of the change

fix #12656 12656

## Brief changelog
add annotation @DubboCrossThread to copy dubbo.tag cross thread like my commit @TraceCrossThread in skywalking(https://github.com/apache/skywalking-java/commit/8612f80260c8dd691a5c8308625fae5fca03ade2),

with @DubboCrossThread , bute-buddy will auto copy dubbo.tag from parent thread to children thread . 

## Verifying this change
new thread run to get dubbo.tag

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Review by songxiaosheng — COMMENTED** [review]
这个包是独立的么，使用这个应该需要强依赖agent吧，没有用skywalking的也可以选择用ttl 需要扩展出来一个一个适配

**Comment by carlvine500** [comment]
> 这个包是独立的么，使用这个应该需要强依赖agent吧

- 静态安装: -javaagent=<agentjar>, 相关的增强代码可以放在agentjar中,业务工程只需要保留注解不引用agent
- 动态安装: ByteBuddyAgent.install(), 增强代码放在业务工程, 需要引用byte-buddy-agent(约250K)

大神有什么建议吗？

**Comment by songxiaosheng** [comment]
> > 这个包是独立的么，使用这个应该需要强依赖agent吧
> 
> - 静态安装: -javaagent= \<agentjar\>, 相关的增强代码可以放在agentjar中,业务工程只需要保留注解不引用agent
> - 动态安装: ByteBuddyAgent.install(), 增强代码放在业务工程, 需要引用byte-buddy-agent(约250K)
> 
> 大神有什么建议吗？ 

学习了

**Comment by carlvine500** [comment]
> 

对性能不敏感的场景使用TTL框架是不错的选择.  
我在增强skywalking的时候也测试过, 性能有一定的损失, 因为它里边把ThreadLocal的内容全部复制到子线程, 当使用ThreadLocal东西较多时, 时间和空间会增加这些消耗,  而使用bute-buddy 增强注解复制指定的内容 ,可以避免复制范围过大 .

**Comment by AlbumenJ** [comment]
Can this be a plugin in `dubbo-spi-extensions` or `dubbo-plugin`? And can we stripping off the strong dependency on spring?

**Comment by carlvine500** [comment]
> 

ok , I will refactor it .

**Comment by carlvine500** [comment]
@AlbumenJ AlbumenJ please review it  .

**Comment by carlvine500** [comment]
@songxiaosheng could you please check if there is any problem ? some checks failing are not about this pull request .

File: dubbo-plugin/dubbo-plugin-cross-thread/src/main/java/org/apache/dubbo/crossthread/interceptor/RunnableOrCallableMethodInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Advice.OnMethodEnter
+    public static void onMethodEnter(
+        @Advice.FieldValue(value = RunnableOrCallableActivation.FIELD_NAME_DUBBO_TAG, readOnly = false) String dubboTag) {
+        // copy tag to RpcContext from RunnableOrCallable's field value
+        RpcContext.getClientAttachment().setAttachment(CommonConstants.TAG_KEY, dubboTag);
+    }
+
+    @Advice.OnMethodExit
+    public static void onMethodExit() {
```
Comment (songxiaosheng) [inline dubbo-plugin/dubbo-plugin-cross-thread/src/main/java/org/apache/dubbo/crossthread/interceptor/RunnableOrCallableMethodInterceptor.java:34]: 这个方法异常的时候会执行么

**Comment by songxiaosheng** [comment]
我还想到一个问题就是这个工具可能只实用在这个agent下面，其他的agent比如skywalking  ttl同时存在也会存在冲突，runnable这些类只能被一个agent重写后面agent重写就会失败，可以文档里声明下

File: dubbo-plugin/dubbo-plugin-cross-thread/src/main/java/org/apache/dubbo/crossthread/interceptor/RunnableOrCallableMethodInterceptor.java (hunk 1 above)
Comment (carlvine500) [reply dubbo-plugin/dubbo-plugin-cross-thread/src/main/java/org/apache/dubbo/crossthread/interceptor/RunnableOrCallableMethodInterceptor.java:34]: 感谢提醒  ! 这里注解少写了, 修改前方法异常不会执行onMethodExit(),修改后会执行:
修改前:`@Advice.OnMethodExit(onThrowable = Throwable.class)`
修改后:`@Advice.OnMethodExit`

**Comment by carlvine500** [comment]
> 我还想到一个问题就是这个工具可能只实用在这个agent下面，其他的agent比如skywalking ttl同时存在也会存在冲突，runnable这些类只能被一个agent重写后面agent重写就会失败，可以文档里声明下

同时使用skywalking ttl确实有冲突 , 文档最前边也加上了会冲突说明 .

**Review by songxiaosheng — APPROVED** [review]
LGTM

**Review by AlbumenJ — CHANGES_REQUESTED** [review]
It would be better to move this implementation to dubbo-spi-extensions

**Comment by carlvine500** [comment]
> It would be better to move this implementation to dubbo-spi-extensions

OK , let me try to refactor .

**Comment by carlvine500** [comment]
@AlbumenJ changes are migrated to  https://github.com/apache/dubbo-spi-extensions/pull/215

**Comment by carlvine500** [comment]
@songxiaosheng 
with bytebuddy's author suggestion [1489](https://github.com/raphw/byte-buddy/issues/1489) , it can run with skywalking and TTL now , 
https://github.com/carlvine500/dubbo-spi-extensions/blob/fee705542ca19dc9e505d4901c098e211403420c/dubbo-cross-thread-extensions/README.md

**Comment by AlbumenJ** [comment]
Let's working on https://github.com/apache/dubbo-spi-extensions/pull/215 ~

**Pull request closed without merging**